File: Graph.scala
import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet
import scala.collection.mutable.ListBuffer
import scala.collection.immutable.List
import util.control.Breaks._
object Graph extends App{
    //Visible to all instances including subclasses
    //private val nodeLookUp:HashMap = HashMap[Int, Node] ()
    // Visible to one instance of  class but including subclasses
    // Use new method if override is used. Apply only works with default setting on companion object
    private[this] val nodeLookUp: HashMap[Int, Node] = new HashMap[Int, Node] (){ override def default(key:Int) = Node(-1,ListBuffer[Node]()) }
    private[this] case class Node(val id:Int, val adjacent:ListBuffer[Node]){}
    private[this] def getNode(id: Int):Node = {
        nodeLookUp(id)
    }
    def addNodes(nodes: List[Int]):Unit = 
        for(node <- nodes) {
            nodeLookUp+=(node->Node(node,ListBuffer[Node]()))
        }
    
    def addEdge(source: Int, destination: Int):Unit = {
        val s:Node = getNode(source)
        val d:Node = getNode(destination)
        s.adjacent+=d
    }
    def hasPathDFS(source:Int, destination: Int):Boolean ={
        val s = getNode(source)
        val d = getNode(destination)
        val visited = HashSet[Int]()
        hasPathDFS(s,d, visited)
    }
    def hasPathBFS(source:Int, destination: Int):Boolean ={
        val s = getNode(source)
        val d = getNode(destination)
        hasPathBFS(s,d)
    }
    private[this] def hasPathBFS(source:Node, destination:Node):Boolean = {
        val nextToVisit = ListBuffer[Node]()
        val visited = HashSet[Int]()
        nextToVisit+=source
        while(!nextToVisit.isEmpty){
            val node = nextToVisit.remove(0)
            if(node == destination)
                return true
            breakable {
                if(visited.contains(node.id))
                    break
                visited+=node.id
                for(child <- node.adjacent)
                    nextToVisit+=child
            }
        }
        return false
    }
    private[this] def hasPathDFS(source:Node, destination:Node,
        visited:HashSet[Int]):Boolean = {
        if(visited.contains(source.id))
            return false
        else{
            visited.add(source.id)
            if(source == destination)
                return true
            for(child <- source.adjacent ){
                if(hasPathDFS(child, destination, visited))
                    return true
            }
        }
        return false
    }
    val vertices = List(0,1,2,3,4,5,6)
    addNodes(vertices)
    addEdge(0,1)
    addEdge(0,2)
    addEdge(1,3)
    addEdge(2,4)
    addEdge(3,5)
    println(nodeLookUp.size)
    println(nodeLookUp)
    println(hasPathDFS(0,6))
    println(hasPathBFS(0,6))
}

File: Graph.py
from collections import deque
class Graph:
    def __init__(self):
        self.__nodeLookUp={}
    class Node:
        def __init__(self,id,adjacent):
            self.__id = id
            self.__adjacent = adjacent
        @property
        def id(self):
            return self.__id
        @property
        def adjacent(self):
            return self.__adjacent
    def addNodes(self,nodes):
        for key in nodes:
            node = self.Node(key,[])
            self.__nodeLookUp[key] = node
    def __getNode(self,id):
        return self.__nodeLookUp[id]
    def addEdge(self,source,destination):
        s = self.__getNode(source)
        d = self.__getNode(destination)
        s.adjacent.append(d)
    def hasPathBFS(self,source, destination):
        s = self.__getNode(source)
        d = self.__getNode(destination)
        return self.__hasPathBFSVisit(s,d)
    def hasPathDFS(self,source, destination):
        s = self.__getNode(source)
        d = self.__getNode(destination)
        visited = set()
        return self.__hasPathDFSVisit(s,d,visited)
    def __hasPathBFSVisit(self,source, destination):
        visited = set()
        l = deque()
        l.append(source)
        while(l):
            node = l.popleft()
            if(node == destination):
                return True
            else:
                if(node.id in visited):
                    continue
                visited.add(node.id)
                for child in node.adjacent:
                    l.append(child)
        return False
    def __hasPathDFSVisit(self,source, destination, visited):
        if(source.id in visited):
            return False
        else:
            visited.add(source.id)
            if(source == destination):
                return True
            for child in source.adjacent:
                if(self.__hasPathDFSVisit(child,destination, visited)):
                    return True
        return False

if __name__ == '__main__':
    g = Graph()
    vertices = [0,1,2,3,4,5,6]
    g.addNodes(vertices)
    g.addEdge(0,1)
    g.addEdge(0,2)
    g.addEdge(1,3)
    g.addEdge(2,4)
    g.addEdge(3,5)
    print(g.hasPathDFS(1,5))
    print(g.hasPathBFS(0,1))
